server/auth.go

```diff
@@ -152,50 +200,91 @@ func (s *Server) checkAuthorization(c *client) bool {
 	}
 }
 
-// hasUsers leyt's us know if we have a users array.
-func (s *Server) hasUsers() bool {
-	s.mu.Lock()
-	hu := s.users != nil
-	s.mu.Unlock()
-	return hu
-}
-
 // isClientAuthorized will check the client against the proper authorization method and data.
-// This could be token or username/password based.
+// This could be nkey, token, or username/password based.
 func (s *Server) isClientAuthorized(c *client) bool {
-	// Snapshot server options.
-	opts := s.getOpts()
+	// Snapshot server options by hand and only grab what we really need.
+	s.optsMu.RLock()
+	customClientAuthentication := s.opts.CustomClientAuthentication
+	authorization := s.opts.Authorization
+	username := s.opts.Username
+	password := s.opts.Password
+	s.optsMu.RUnlock()
+
+	// Check custom auth first, then nkeys, then multiple users, then token, then single user/pass.
+	if customClientAuthentication != nil {
+		return customClientAuthentication.Check(c)
+	}
 
-	// Check custom auth first, then multiple users, then token, then single user/pass.
-	if opts.CustomClientAuthentication != nil {
-		return opts.CustomClientAuthentication.Check(c)
-	} else if s.hasUsers() {
-		s.mu.Lock()
-		user, ok := s.users[c.opts.Username]
+	var nkey *NkeyUser
+	var user *User
+	var ok bool
+
+	s.mu.Lock()
+	authRequired := s.info.AuthRequired
+	if !authRequired {
+		// TODO(dlc) - If they send us credentials should we fail?
 		s.mu.Unlock()
+		return true
+	}
 
+	// Check if we have nkeys or users for client.
+	hasNkeys := s.nkeys != nil
+	hasUsers := s.users != nil
+	if hasNkeys && c.opts.Nkey != "" {
+		nkey, ok = s.nkeys[c.opts.Nkey]
+		if !ok {
+			s.mu.Unlock()
+			return false
+		}
+	} else if hasUsers && c.opts.Username != "" {
+		user, ok = s.users[c.opts.Username]
 		if !ok {
+			s.mu.Unlock()
+			return false
+		}
+	}
+	s.mu.Unlock()
+
+	// Verify the signature against the nonce.
+	if nkey != nil {
+		if c.opts.Sig == "" {
 			return false
 		}
+		sig, err := base64.RawURLEncoding.DecodeString(c.opts.Sig)
```

Comment:
This needs to be StdEncoding. Client libraries that encode to base64 will emit the padding by default.

Reply:
I think we document that we require the URLEncoding version which should be easily present in all libs since that is the main use of it IMO>

Reply:
that is exactly the issue. We need the one with padding because that is the one that is prevalent.

Reply:
Not sure I agree. Both are common, the one we use is required for inclusion in any url. So auth tokens etc would need to be in this format I believe.